### PR TITLE
Add "mod" items to the structure view

### LINKED ISSUE #47: Improve file structure view
File structure (aka imenu) lists things, declared in a file. It is accessible via `Alt-7` or `Ctrl-F12`.

Currently it only supports a flat list of functions and methods. It should support hierarchical view of `struct`s, `trait`s, `mod`s and `impl`s.

Relevant code is here: https://github.com/alexeykudinkin/intellij-rust/tree/master/src/org/rust/lang/structure

### PR DESCRIPTION
**Pull request by Turbo87**
bildschirmfoto 2015-11-06 um 13 17 07

resolves #47

### CODE REVIEW

File: src/org/rust/lang/structure/RustModTreeElement.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import com.intellij.ide.structureView.StructureViewTreeElement
+import com.intellij.ide.structureView.impl.common.PsiTreeElementBase
+import org.rust.lang.core.psi.*
+
+class RustModTreeElement(file: RustModItem) : PsiTreeElementBase<RustModItem>(file) {
+    override fun getPresentableText() = element?.name
+
+    override fun getChildrenBase(): Collection<StructureViewTreeElement> {
+        return element?.getDeclarations()?.map { toTreeElement(it) }?.filterNotNull().orEmpty()
```
Comment (matklad) [inline src/org/rust/lang/structure/RustModTreeElement.kt]: I think it's better to use `element.items` here:

* it preserves more type information
* `getDeclarations` might go away soon

File: src/org/rust/lang/structure/RustModTreeElement.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package org.rust.lang.structure
+
+import com.intellij.ide.structureView.StructureViewTreeElement
+import com.intellij.ide.structureView.impl.common.PsiTreeElementBase
+import org.rust.lang.core.psi.*
+
+class RustModTreeElement(file: RustModItem) : PsiTreeElementBase<RustModItem>(file) {
```
Comment (matklad) [inline src/org/rust/lang/structure/RustModTreeElement.kt]: Just a cosmetics: may be `mod: RustModItem` is better?

Comment (Turbo87) [inline src/org/rust/lang/structure/RustModTreeElement.kt]: sorry, yeah, copy-paste problem...

**Comment by Turbo87** [comment]
@matklad fixed

**Comment by matklad** [comment]
A folder icon looks misleading: a `mod` is a file. But I guess we have no better choice at the moment.

**Comment by Turbo87** [comment]
yeah, I think we should just reuse the existing icons for now until we can think of something better.

@matklad could you find out how the intellij icons are actually created at jetbrains? are there some sort of SVG files to generate them? if so we could probably modify them to keep the look and feel but properly add things like `impl` blocks etc.

**Pull request merged**